docs/data/data-grid/recipes-editing/BulkEditing.tsx

```diff
@@ -0,0 +1,206 @@
+
+export default function BulkEditing() {
+  const { data } = useDemoData({
+    dataSet: 'Commodity',
+    rowLength: 100,
+    maxColumns: 7,
+    editable: true,
+    visibleFields: [
+      'id',
+      'commodity',
+      'traderName',
+      'traderEmail',
+      'quantity',
+      'filledQuantity',
+    ],
+  });
+
+  const apiRef = useGridApiRef();
+
+  const [hasUnsavedRows, setHasUnsavedRows] = React.useState(false);
+  const unsavedChangesRef = React.useRef<{
+    unsavedRows: Record<GridRowId, GridValidRowModel>;
+    rowsBeforeChange: Record<GridRowId, GridValidRowModel>;
+  }>({
+    unsavedRows: {},
+    rowsBeforeChange: {},
+  });
+  const [isSaving, setIsSaving] = React.useState(false);
+
+  const columns = React.useMemo<GridColDef[]>(() => {
+    return [
+      {
+        field: 'actions',
+        type: 'actions',
+        getActions: ({ id, row }) => {
+          return [
+            <GridActionsCellItem
+              icon={<RestoreIcon />}
+              label="Discard changes"
+              disabled={unsavedChangesRef.current.unsavedRows[id] === undefined}
+              onClick={() => {
+                apiRef.current.updateRows([
+                  unsavedChangesRef.current.rowsBeforeChange[id],
+                ]);
+                delete unsavedChangesRef.current.rowsBeforeChange[id];
+                delete unsavedChangesRef.current.unsavedRows[id];
+                setHasUnsavedRows(
+                  Object.keys(unsavedChangesRef.current.unsavedRows).length > 0,
+                );
+              }}
+            />,
+            <GridActionsCellItem
+              icon={<DeleteIcon />}
+              label="Delete"
+              onClick={() => {
+                unsavedChangesRef.current.unsavedRows[id] = {
+                  ...row,
+                  _action: 'delete',
+                };
+                if (!unsavedChangesRef.current.rowsBeforeChange[id]) {
+                  unsavedChangesRef.current.rowsBeforeChange[id] = row;
+                }
+                setHasUnsavedRows(true);
+                apiRef.current.updateRows([row]); // to trigger row render
+              }}
+            />,
+          ];
+        },
+      },
+      ...data.columns,
+    ];
+  }, [data.columns, unsavedChangesRef, apiRef]);
+
+  const processRowUpdate: NonNullable<DataGridProps['processRowUpdate']> = (
+    newRow,
+    oldRow,
+  ) => {
+    const rowId = newRow.id;
+
+    unsavedChangesRef.current.unsavedRows[rowId] = newRow;
+    if (!unsavedChangesRef.current.rowsBeforeChange[rowId]) {
+      unsavedChangesRef.current.rowsBeforeChange[rowId] = oldRow;
+    }
+    setHasUnsavedRows(true);
+    return newRow;
+  };
+
+  const discardChanges = () => {
+    setHasUnsavedRows(false);
+    Object.values(unsavedChangesRef.current.rowsBeforeChange).forEach((row) => {
+      apiRef.current.updateRows([row]);
+    });
+    unsavedChangesRef.current = {
+      unsavedRows: {},
+      rowsBeforeChange: {},
+    };
+  };
```

Comment:
not sure, but shouldn't this be in a `useCallback` because of the `apiRef`?

Reply:
It could be a `useCallback` so it's only created once (since all dependencies are stable) 👍🏻